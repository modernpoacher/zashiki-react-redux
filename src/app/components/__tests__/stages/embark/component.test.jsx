import React from 'react'
import renderer from 'react-test-renderer'

import {
  RESOLVED,
  REJECTED,
  PENDING
} from '#zashiki-react-redux/app/common'

import Embark from '#zashiki-react-redux/app/components/stages/embark/component'

jest.mock('#zashiki-react-redux/app/common', () => ({
  RESOLVED: 'MOCK RESOLVED',
  REJECTED: 'MOCK REJECTED',
  PENDING: 'MOCK PENDING'
}))

jest.mock('#zashiki-react-redux/app/components/stages/embark/status/resolved', () => () => 'MOCK RESOLVED')
jest.mock('#zashiki-react-redux/app/components/stages/embark/status/rejected', () => () => 'MOCK REJECTED')
jest.mock('#zashiki-react-redux/app/components/stages/embark/status/pending', () => () => 'MOCK PENDING')

jest.mock('#zashiki-react-redux/app/router/with-router', () => (Component) => Component)

describe('#zashiki-react-redux/app/components/stages/embark/component', () => {
  const MOCK_DEFINITION = {
    meta: {
      uri: '#/'
    },
    elements: {
      title: 'MOCK TITLE',
      field: {
        id: 'MOCK ID'
      }
    }
  }

  const MOCK_RESOURCE = {
    alpha: 'MOCK ALPHA',
    omega: 'MOCK OMEGA'
  }

  const MOCK_RESPONSE = {
    '#/': 'MOCK VALUE'
  }

  const MOCK_ERRORS = []
  const MOCK_ONCHANGE = jest.fn()
  const MOCK_ONSUBMIT = jest.fn()
  const MOCK_ONEMBARK = jest.fn()

  describe('Always', () => {
    const component = (
      <Embark
        status={RESOLVED}
        definition={MOCK_DEFINITION}
        resource={MOCK_RESOURCE}
        response={MOCK_RESPONSE}
        errors={MOCK_ERRORS}
        onChange={MOCK_ONCHANGE}
        onSubmit={MOCK_ONSUBMIT}
        onEmbark={MOCK_ONEMBARK}
      />
    )

    const spy = jest.spyOn(Embark.prototype, 'componentDidMount')

    let rendered

    beforeEach(() => {
      rendered = renderer.create(component)
    })

    describe('Always', () => {
      it('renders', () => {
        expect(rendered.toJSON())
          .toMatchSnapshot()
      })
    })

    it('invokes `componentDidMount`', () => {
      expect(spy)
        .toHaveBeenCalled()
    })

    it('invokes `onEmbark`', () => {
      expect(MOCK_ONEMBARK)
        .toHaveBeenCalled()
    })
  })

  describe('`RESOLVED`', () => {
    it('renders', () => {
      const component = (
        <Embark
          status={RESOLVED}
          definition={MOCK_DEFINITION}
          resource={MOCK_RESOURCE}
          response={MOCK_RESPONSE}
          errors={MOCK_ERRORS}
          onChange={MOCK_ONCHANGE}
          onSubmit={MOCK_ONSUBMIT}
          onEmbark={MOCK_ONEMBARK}
        />
      )

      expect(renderer.create(component).toJSON())
        .toMatchSnapshot()
    })
  })

  describe('`REJECTED`', () => {
    it('renders', () => {
      const component = (
        <Embark
          status={REJECTED}
          definition={MOCK_DEFINITION}
          resource={MOCK_RESOURCE}
          response={MOCK_RESPONSE}
          errors={MOCK_ERRORS}
          exception={{
            name: 'MOCK NAME',
            message: 'MOCK MESSAGE'
          }}
          onChange={MOCK_ONCHANGE}
          onSubmit={MOCK_ONSUBMIT}
          onEmbark={MOCK_ONEMBARK}
        />
      )

      expect(renderer.create(component).toJSON())
        .toMatchSnapshot()
    })
  })

  describe('`PENDING`', () => {
    it('renders', () => {
      const component = (
        <Embark
          status={PENDING}
          definition={MOCK_DEFINITION}
          onChange={MOCK_ONCHANGE}
          onSubmit={MOCK_ONSUBMIT}
          onEmbark={MOCK_ONEMBARK}
        />
      )

      expect(renderer.create(component).toJSON())
        .toMatchSnapshot()
    })
  })
})
