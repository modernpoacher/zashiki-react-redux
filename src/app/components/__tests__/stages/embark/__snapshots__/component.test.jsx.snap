// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#zashiki-react-redux/app/components/stages/embark/component \`PENDING\` renders 1`] = `"MOCK PENDING"`;

exports[`#zashiki-react-redux/app/components/stages/embark/component \`REJECTED\` renders 1`] = `"MOCK REJECTED"`;

exports[`#zashiki-react-redux/app/components/stages/embark/component \`RESOLVED\` renders 1`] = `"MOCK RESOLVED"`;

exports[`#zashiki-react-redux/app/components/stages/embark/component Always Always renders 1`] = `"MOCK RESOLVED"`;
